libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -101,29 +102,53 @@ void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(ContractDefinitio
 	}
 
 	auto const* baseSlotExpressionType = type(baseSlotExpression);
-	auto const* rationalType = dynamic_cast<RationalNumberType const*>(baseSlotExpressionType);
-	if (!rationalType)
+	if (
+		!dynamic_cast<IntegerType const*>(baseSlotExpressionType) &&
+		!dynamic_cast<RationalNumberType const*>(baseSlotExpressionType)
+	)
 	{
 		m_errorReporter.typeError(
 			6396_error,
 			baseSlotExpression.location(),
-			"The base slot of the storage layout must evaluate to a rational number."
+			"The base slot of the storage layout must evaluate to an integer number."
```

Comment:
```suggestion
			"The base slot of the storage layout must evaluate to an integer."
```

An integer is a number, so this is a bit redundant. Also, couldn't you reuse error 1763? It's the same error and provides the same source location.

Reply:
The concern I'd have with reusing error `1763` is that although it has the same location and message (which we could try to improve in both cases to differentiate them), it is a different case where the expression is a rational but it is fractional. In the case here of error `6396` the expression is neither a rational nor an integer number.
We can change the message here to `The base slot of the storage layout must evaluate to a number.` to clarify and distinguish the errors. What do you think?

Reply:
I think it's fine to cover both cases with the same error. The exact case is different (fractional literal vs non-integer constant), but conceptually the issue is the same with both - the user is trying to use something that's not an integer and we can't use that as a slot number.

Reply:
Though, looking at the test cases, perhaps some clarification would be useful in cases where user is trying to use something like `address` or `bytesNN` since those can be initialized with an integer.

Reply:
Should we then change the message here to `The base slot type must be a numerical type` or something similar?

Reply:
I think I'd keep it as "must evaluate to an integer" and just do a single check. It's hard to phrase these things.

What I'd really want to do eventually would be to rewrite the docs to introduce clear terminology.

Reply:
I added a check to append a hint in the cases where the expression is of type `address` or `bytesN`.